image-classifier/index.ts

```diff
@@ -0,0 +1,165 @@
+
+import { ImageClassifier, FilesetResolver } from '@mediapipe/tasks-vision';
+
+// Get DOM elements
+const video = document.getElementById('webcam') as HTMLVideoElement;
+const webcamPredictions = document.getElementById('webcamPredictions');
+const demosSection = document.getElementById('demos');
+let enableWebcamButton: HTMLButtonElement;
+let webcamRunning: Boolean = false;
+const videoHeight = '360px';
+const videoWidth = '480px';
+
+const imageContainers = document.getElementsByClassName('classifyOnClick');
+let runningMode: 'image' | 'video' = 'image';
+
+// Add click event listeners for the img elements.
+for (let i = 0; i < imageContainers.length; i++) {
+  imageContainers[i].children[0].addEventListener('click', handleClick);
+}
+
+// Track imageClassifier object and load status.
+let imageClassifier: ImageClassifier;
+
+/**
+ * Demo 1: Classify images on click and display results.
+ */
+async function handleClick(event) {
+  // Do not classify if imageClassifier hasn't loaded
+  if (imageClassifier === undefined) {
+    return;
+  }
+  // if video mode is initialized, set runningMode to image
+  if (runningMode === 'video') {
+    runningMode = 'image';
+    await imageClassifier.setOptions({ runningMode });
+  }
+
+  // imageClassifier.classify() returns a promise which, when resolved, is a ClassificationResult object.
+  // Use the ClassificationResult to print out the results of the prediction.
+  const classificationResult = await imageClassifier.classify(event.target);
+  // Write the predictions to a new paragraph element and add it to the DOM.
+  const classifications = classificationResult.classifications;
+
+  const p = event.target.parentNode.childNodes[3];
+  p.className = 'classification';
+  p.innerText = `Classification: ${classifications[0].categories[0].categoryName}
+                 Confidence: ${Math.round(parseFloat(`${classifications[0].categories[0].score}`) * 100)}%`;
+    // 'Classificaton: ' +
+    // classifications[0].categories[0].categoryName +
+    // '\n Confidence: ' +
+    // Math.round(parseFloat(`${classifications[0].categories[0].score}`) * 100) +
+    // '%';
+}
+
+/********************************************************************
+// Demo 2: Continuously grab image from webcam stream and classify it.
+********************************************************************/
+
+// Check if webcam access is supported.
+function hasGetUserMedia() {
+  return !!(navigator.mediaDevices && navigator.mediaDevices.getUserMedia);
+}
+
+// Get classification from the webcam
+async function predictWebcam() {
+  // Do not classify if imageClassifier hasn't loaded
+  if (imageClassifier === undefined) {
+    return;
+  }
+  // if image mode is initialized, create a new classifier with video runningMode
+  if (runningMode === 'image') {
+    runningMode = 'video';
+    // await runDemo();
+    await imageClassifier.setOptions({ runningMode: runningMode });
+  }
+  let nowInMs = Date.now();
```

Comment:
Where we can, we should use performance.now() instead of Date.now().

Reasons:
- only the former is monotonically increasing (otherwise can cause crashes)
- former can potentially be more accurate
- we don't need absolute time reference, so more clear to readers of code